lib/addressable/uri.rb

```diff
@@ -1216,7 +1216,34 @@ def tld=(new_tld)
     # @example
     #   Addressable::URI.parse("http://www.example.co.uk").domain # => "example.co.uk"
     def domain
-      PublicSuffix.domain(self.host, ignore_private: true)
+      public_suffix.domain
+    end
+
+    ##
+    # Returns the first subdomain for this host.
+    #
+    # @example
+    #   Addressable::URI.parse("http://www.test.example.co.uk").domain # => "www"
+    def subdomain
+      subdomains.first
```

Comment:
Still think there's ambiguity around what's a [subdomain](https://en.wikipedia.org/wiki/Subdomain) and what a method like this is expected to return. 

`www` is a subdomain of `test.example.co.uk`
`test` is a subdomain of `example.co.uk`
`example` is a subdomain of `co.uk`

This method would return `nil` when parsing `http://example.co.uk`

Reply:
I'm not familiar with that in Rails, can you link the source for method? When is it typically used?

I'm not sure what we want here, if the method should exist?

Reply:
Multi-tenant applications or anything that uses subdomains (think Shopify) use this. I've always been surprised there's no way to get the subdomain(s) from URIs in Ruby. Rails will do it for the current request, but that's it and it's not Public Suffix aware. You can only specify a single TLD for your entire Rails app, so you need to reach for something else like addressable + public_suffix.

https://api.rubyonrails.org/classes/ActionDispatch/Http/URL.html#method-i-subdomain

Reply:
Any thoughts on this?

Reply:
Yeah... _if_ we should add this method, is it not better if it matches the `subdomain` on ActionDispatch? From the URL above
> Returns all the subdomains as a string, so "dev.www" would be returned for “dev.www.rubyonrails.org”. 

For `http://www.test.example.co.uk` that would be `www.test`, not just `www`?

Reply:
What about not adding the `#subdomain` method? Instead only `#subdomains`. It is easy for the user to do `subdomains.first` or `subdomains.first(2)` etc. depending on what they want.